README.md
# **rn-tween**

It can help animating components by timing.

## Installation

- Using **Npm**

```bash
npm install rn-tween --save
```

- Using **Yarn**

```bash
yarn add rn-tween
```

## Tween components

- RNTween
- RNTweenView
- RNTweenImage
- RNTweenText

## Example

### Auto start

```jsx
import {RNTweenView} from 'rn-tween';

<RNTweenView
  autoStartName="transit"
  initialConfig={{
    transit: {
      mode: 'parallel',
      configs: [
        {
          from: 0,
          to: 1,
          duration: 1000,
          styleProperty: 'opacity',
          loop: true,
        },
        {
          from: 0.2,
          to: 1,
          duration: 1000,
          styleProperty: 'scale',
          loop: true,
        },
      ],
    },
  }}
  style={{
    width: 100,
    height: 100,
    backgroundColor: 'red',
  }}
/>;
```

### Call functions

```jsx
import {RNTweenView} from 'rn-tween';

<RNTweenView
  ref={ref => (_tween = ref)}
  firstUsedConfigName="transit"
  initialConfig={{
    transit: {
      mode: 'parallel',
      configs: [
        {
          from: 0,
          to: 1,
          duration: 500,
          styleProperty: 'opacity',
        },
        {
          from: 0.2,
          to: 1,
          duration: 500,
          styleProperty: 'scale',
        },
      ],
    },
  }}
  style={{
    width: 100,
    height: 100,
    backgroundColor: 'orange',
  }}
/>;

// Stop animating the tweeen with a specific configuration name
_tween.start({name: 'transit'});
_tween.start({name: 'transit', reversed: true});
_tween.start({name: 'transit', reversed: true, onComplete: () => alert('Finished')});

// Stop animating the tween
_tween.stop();

// Reprepare tween animation configurations
_tween.prepare({
  transit: {
    mode: 'sequence',
    configs: [
      {
        from: 0,
        to: 1,
        duration: 500,
        styleProperty: 'opacity',
      },
      {
        from: 0.2,
        to: 1,
        duration: 500,
        styleProperty: 'scale',
      },
    ],
  },
});
```

## Available component props

## RNTween

| Name                | Type      | Default | Description                                                                  |
| ------------------- | --------- | ------- | ---------------------------------------------------------------------------- |
| autoStartName       | string    | null    | Tell which animation tween should automatically start at the first time      |
| firstUsedConfigName | string    | null    | Tell which animation tween configuration should be applied at the first time |
| initialConfig       | object    | null    | The initial tween configurations to be used                                  |
| AnimatedComponent   | Component | null    | The animated component to be animated                                        |
| onComplete          | function  | null    | The callback invoked after the tween animation completed                     |

## RNTweenView

Also inherits [ViewProps](https://facebook.github.io/react-native/docs/view)

| Name                | Type     | Default | Description                                                                  |
| ------------------- | -------- | ------- | ---------------------------------------------------------------------------- |
| autoStartName       | string   | null    | Tell which animation tween should automatically start at the first time      |
| firstUsedConfigName | string   | null    | Tell which animation tween configuration should be applied at the first time |
| initialConfig       | object   | null    | The initial tween configurations to be used                                  |
| onComplete          | function | null    | The callback invoked after the tween animation completed                     |

## RNTweenText

Also inherits [TextProps](https://facebook.github.io/react-native/docs/text)

| Name                | Type     | Default | Description                                                                  |
| ------------------- | -------- | ------- | ---------------------------------------------------------------------------- |
| autoStartName       | string   | null    | Tell which animation tween should automatically start at the first time      |
| firstUsedConfigName | string   | null    | Tell which animation tween configuration should be applied at the first time |
| initialConfig       | object   | null    | The initial tween configurations to be used                                  |
| onComplete          | function | null    | The callback invoked after the tween animation completed                     |

## RNTweenImage

Also inherits [ImageProps](https://facebook.github.io/react-native/docs/image)

| Name                | Type     | Default | Description                                                                  |
| ------------------- | -------- | ------- | ---------------------------------------------------------------------------- |
| autoStartName       | string   | null    | Tell which animation tween should automatically start at the first time      |
| firstUsedConfigName | string   | null    | Tell which animation tween configuration should be applied at the first time |
| initialConfig       | object   | null    | The initial tween configurations to be used                                  |
| onComplete          | function | null    | The callback invoked after the tween animation completed                     |

## initialConfig

```jsx
initialConfig: {
  [name: string]: {
    mode,
    configs
  }
}
```

| Name    | Type   | Description                                                    |
| ------- | ------ | -------------------------------------------------------------- |
| mode    | string | Specify tween animation mode like **parallel** or **sequence** |
| configs | array  | Array of tween animation configuration                         |

### initialConfig[name].configs[index]

| Name          | Type   | Description                                                                |
| ------------- | ------ | -------------------------------------------------------------------------- |
| styleProperty | string | Specify any style property to be animated like **opacity**, **scale**, etc |
| from          | number | Value from where the animation starts                                      |
| to            | number | Value to where the animation must reach                                    |
| duration      | number | Duration of animation                                                      |
| delay         | number | Delay before the animation starts                                          |
| loop          | bool   | Tell if the animation should play as loop                                  |
| easing        | function | Easing function to define curve. Default is TweenEasing.linear           |
| useNative     | bool   | Uses the native driver when true. Default is true                          |

## Available instance functions

### start({ name, reversed, onComplete })

| Name       | Type     | Description                                   |
| ---------- | -------- | --------------------------------------------- |
| name       | string   | Name of tween animation configuration         |
| reversed   | bool     | Tell if the animation should play in reverse  |
| onComplete | function | Callback invoked after the animation complete |

### stop()

Stop a running animation

### prepare({ mode, configs })

| Name    | Type   | Description                                                    |
| ------- | ------ | -------------------------------------------------------------- |
| mode    | string | Specify tween animation mode like **parallel** or **sequence** |
| configs | array  | Array of tween animation configuration                         |

## License

- [MIT](LICENSE)


CHANGELOG.md
# 1.0.4 (2019-08-29)

- Added property **easing** and **useNative** for the tween configuration

# 1.0.2 (2019-08-28)

- Fixed issue that children of tween components are not displayed

# 1.0.1 (2019-08-28)

- Initial release

src/RNTween.js
import React from 'react';
import { Animated, Easing } from 'react-native';

const transformProperties = [
	'perspective',
	'rotate',
	'rotateX',
	'rotateY',
	'rotateZ',
	'scale',
	'scaleX',
	'scaleY',
	'skewX',
	'skewY',
	'translateX',
	'translateY'
];

export { Easing as TweenEasing };

const defaultProps = {
	autoStartName: null,
	firstUsedConfigName: null,
	initialConfig: null,
	onComplete: null
};

export class RNTween extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			animatedStyles: [],
			tweenConfig: {},
			playedTweenName: '',
			mounted: false
		};

		this.prepare(props.initialConfig);

		const configNames = Object.keys(this.state.tweenConfig);
		if (configNames.length > 0) {
			const config = this.state.tweenConfig[
				props.firstUsedConfigName || configNames[0]
			];

			if (config) {
				this.state.animatedStyles = this._getComputedAnimatedStyles(config);
			}
		}

		this.props.autoStartName &&
			this.start({ name: this.props.autoStartName, reversed: false });
	}

	componentDidMount() {
		this.setState({ mounted: true });
	}

	render() {
		const { style, AnimatedComponent, ...restProps } = this.props;
		return (
			<AnimatedComponent
				{...restProps}
				style={[style, ...this.state.animatedStyles]}
			/>
		);
	}

	_createAnimatedTween(config, reversed) {
		const {
			delay,
			value,
			from,
			to,
			interpolated,
			duration,
			easing,
			useNative
		} = config;

		value.setValue(
			reversed ? (interpolated ? 1 : to) : interpolated ? 0 : from
		);

		const createTiminingTween = () =>
			Animated.timing(value, {
				toValue: reversed ? (interpolated ? 0 : from) : interpolated ? 1 : to,
				delay: delay === null || delay === undefined ? 0 : delay,
				duration: duration === null || duration === undefined ? 0 : duration,
				useNativeDriver:
					useNative === null || useNative === undefined ? true : useNative,
				easing: easing || Easing.linear
			});

		if (config.loop) {
			return Animated.loop(createTiminingTween());
		} else {
			return createTiminingTween();
		}
	}

	_getComputedAnimatedStyles(config) {
		const animatedStyles = config.configs
			.filter(e => (e.styleProperty ? true : false))
			.map(e => {
				const isStyleForTransform = transformProperties.includes(
					e.styleProperty
				);

				let value = null;

				if (e.interpolated) {
					value = e.value.interpolate({
						inputRange: [0, 1],
						outputRange: [e.from, e.to]
					});
				} else {
					value = e.value;
				}

				if (isStyleForTransform) {
					return {
						transform: [
							{
								[e.styleProperty]: value
							}
						]
					};
				} else {
					return {
						[e.styleProperty]: value
					};
				}
			});
		return animatedStyles;
	}

	prepare(config) {
		this.stop();

		if (!config) {
			return;
		}

		const tweenConfig = {};

		for (let k in config) {
			const c = config[k];
			const newConfig = { mode: c.mode };

			newConfig.configs = c.configs
				.filter(e => (e.styleProperty ? true : false))
				.map(e => {
					return {
						...e,
						value: new Animated.Value(typeof e.from === 'string' ? 0 : e.from),
						interpolated: typeof e.from === 'string'
					};
				});

			tweenConfig[k] = newConfig;
		}

		if (this.state.mounted) {
			this.setState({ tweenConfig });
		} else {
			this.state = { ...this.state, tweenConfig };
		}
	}

	start(option) {
		const config = this.state.tweenConfig[option.name];

		if (!config) {
			return;
		}

		this.stop();

		if (this.state.playedTweenName != option.name) {
			const animatedStyles = this._getComputedAnimatedStyles(config);

			if (this.state.mounted) {
				this.setState({
					playedTweenName: option.name,
					animatedStyles
				});
			} else {
				this.state = {
					...this.state,
					playedTweenName: option.name,
					animatedStyles
				};
			}
		}

		const tweens = config.configs.map(e =>
			this._createAnimatedTween(e, option.reversed)
		);

		let tweenStarter =
			!config.mode || config.mode == 'parallel'
				? Animated.parallel
				: Animated.sequence;

		this._runningTween = tweenStarter(tweens);
		this._runningTween.start(() => {
			this._runningTween = null;
			this.props.onComplete && this.props.onComplete();
			option.onComplete && option.onComplete();
		});
	}

	stop() {
		this._runningTween && this._runningTween.stop();
		this._runningTween = null;
	}
}

RNTween.defaultProps = {
	...defaultProps
};

export class RNTweenView extends React.Component {
	render() {
		const { AnimatedComponent, ...restProps } = this.props;
		return (
			<RNTween
				ref={this._onRef}
				{...restProps}
				AnimatedComponent={Animated.View}
			/>
		);
	}

	_onRef = ref => {
		this._tween = ref;
	};

	prepare(config) {
		this._tween && this._tween.prepare(config);
	}

	start(option) {
		this._tween && this._tween.start(option);
	}

	stop() {
		this._tween && this._tween.stop();
	}
}

RNTweenView.defaultProps = {
	...defaultProps
};

export class RNTweenText extends React.Component {
	render() {
		const { AnimatedComponent, ...restProps } = this.props;
		return (
			<RNTween
				ref={this._onRef}
				{...restProps}
				AnimatedComponent={Animated.Text}
			/>
		);
	}

	_onRef = ref => {
		this._tween = ref;
	};

	prepare(config) {
		this._tween && this._tween.prepare(config);
	}

	start(option) {
		this._tween && this._tween.start(option);
	}

	stop() {
		this._tween && this._tween.stop();
	}
}

RNTweenText.defaultProps = {
	...defaultProps
};

export class RNTweenImage extends React.Component {
	render() {
		const { AnimatedComponent, ...restProps } = this.props;
		return (
			<RNTween
				ref={this._onRef}
				{...restProps}
				AnimatedComponent={Animated.Image}
			/>
		);
	}

	_onRef = ref => {
		this._tween = ref;
	};

	prepare(config) {
		this._tween && this._tween.prepare(config);
	}

	start(option) {
		this._tween && this._tween.start(option);
	}

	stop() {
		this._tween && this._tween.stop();
	}
}

RNTweenImage.defaultProps = {
	...defaultProps
};


src/RNTween.d.ts
import { Component, ComponentType } from 'react';
import { ViewProps, TextProps, ImageProps, EasingFunction, Easing } from 'react-native';

interface RNTweenConfig {
    [index: string]: {
        mode?: 'parallel' | 'sequence' | null,
        configs?: {
            styleProperty: string,
            from: number | string,
            to: number | string,
            duration?: number,
            delay?: number,
            loop?: boolean | null,
            easing?: EasingFunction | null,
            useNative?: boolean | null
        }[]
    };
}

interface RNTweenProps {
    autoStartName?: string | null;
    firstUsedConfigName?: string | null;
    initialConfig?: RNTweenConfig | null;
    AnimatedComponent?: ComponentType<any> | null;
    onComplete?: () => void | null;
}

export {
    Easing as TweenEasing
};

export class RNTween extends Component<RNTweenProps> {
    prepare(config: RNTweenConfig | null): void;

    start(option: {
        name: string,
        reversed: boolean,
        onComplete: () => void
    }): void;

    stop(): void;
}


interface RNTweenViewProps extends ViewProps, RNTweenProps {
}

export class RNTweenView extends Component<RNTweenViewProps> {
    prepare(config: RNTweenConfig | null): void;

    start(option: {
        name: string,
        reversed: boolean,
        onComplete: () => void
    }): void;

    stop(): void;
}

interface RNTweenTextProps extends TextProps, RNTweenProps {
}

export class RNTweenText extends Component<RNTweenTextProps> {
    prepare(config: RNTweenConfig | null): void;

    start(option: {
        name: string,
        reversed: boolean,
        onComplete: () => void
    }): void;

    stop(): void;
}

interface RNTweenImageProps extends ImageProps, RNTweenProps {
}

export class RNTweenImage extends Component<RNTweenImageProps> {
    prepare(config: RNTweenConfig | null): void;

    start(option: {
        name: string,
        reversed: boolean,
        onComplete: () => void
    }): void;

    stop(): void;
}

examples/Example.js
import React from 'react';
import { StyleSheet, Button, SafeAreaView } from 'react-native';
import { RNTweenView, TweenEasing } from '../src/RNTween';

export default class Example extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			isPlaying: false
		};
	}

	render() {
		return (
			<SafeAreaView style={styles.container}>
				<RNTweenView
					autoStartName='transit1'
					initialConfig={{
						transit1: {
							mode: 'parallel',
							configs: [
								{
									from: 0,
									to: 1,
									duration: 1000,
									styleProperty: 'opacity',
									loop: true,
									useNative: true
								},
								{
									from: 0.2,
									to: 1,
									duration: 1000,
									styleProperty: 'scale',
									loop: true,
									useNative: true,
									easing: TweenEasing.bounce
								}
							]
						}
					}}
					style={{
						width: 100,
						height: 100,
						borderRadius: 50,
						backgroundColor: 'red',
						alignSelf: 'center'
					}}
				/>

				<RNTweenView
					ref={ref => (this._tween = ref)}
					firstUsedConfigName='transit2'
					initialConfig={{
						transit2: {
							mode: 'parallel',
							configs: [
								{
									from: 0,
									to: 1,
									duration: 500,
									styleProperty: 'opacity'
								},
								{
									from: 0.2,
									to: 1,
									duration: 500,
									styleProperty: 'scale'
								}
							]
						}
					}}
					style={{
						marginVertical: 50,
						width: 100,
						height: 100,
						backgroundColor: 'orange',
						alignSelf: 'center'
					}}
				/>

				<Button
					title={this.state.isPlaying ? 'Stop' : 'Start'}
					onPress={() => {
						this.setState({ isPlaying: !this.state.isPlaying }, () => {
							if (this.state.isPlaying) {
								this._tween.start({
									name: 'transit2',
									onComplete: () => {
										this.setState({ isPlaying: false });
									}
								});
							} else {
								this._tween.stop();
							}
						});
					}}
				/>
			</SafeAreaView>
		);
	}
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: '#fff',
		padding: 10
	}
});
